cql3/statements/select_statement.cc

```diff
@@ -462,7 +462,13 @@ select_statement::execute_without_checking_exception_message_aggregate_or_paged(
             state, options, command, std::move(key_ranges), _restrictions_need_filtering ? _restrictions : nullptr);
 
     if (aggregate || nonpaged_filtering) {
-        auto builder = cql3::selection::result_set_builder(*_selection, now, *_group_by_cell_indices);
+        std::optional<uint64_t> limit;
+        if (_limit) {
+            auto x = expr::evaluate(*_limit, options);
+            auto l = x.view().validate_and_deserialize<int32_t>(*int32_type);
```

Comment:
`limit` is defined as `uint64_t` but is being deserialized to `int32_t`, is that ok?
I mean, I don't expect negative values in limit, I also don't expect it to be such a huge number to cause an overflow, but one, you never know, two, sanitizers may fail this piece of code.

Reply:
Fail the query on a negative limit. Perhaps we should also fail on limit = 0.

Reply:
```
avi=# select * from t1 limit 2;
 x | y 
---+---
 1 |  
 5 | 4
(2 rows)

avi=# select * from t1 limit -1;
ERROR:  LIMIT must not be negative
avi=# select * from t1 limit 0;
 x | y 
---+---
(0 rows)
```

Reply:
There's a get_limit() function that does this correctly.

Reply:
and, in fact, do_execute() already calls it.

Reply:
Exactly, I mentioned select_statement::get_limit() in my comment above, and it does all the right conversions and checks. I also said I assume we already call it, so I wonder why we found ourselves needing to call it again.

Reply:
What do you mean it doesn't yield the limit value? What does it return? It seems you copied most of its code already (and then forgot you copied the right type ;-)), so it should work...

Reply:
btw, what happens when paging? If the limit is N and we return R<N rows, the new limit should be N-R.

query_pager deals with it, but the logic is so complicated, I wasn't able to make sense of it.

Reply:
I see no generic reason why it would be unacceptable. It seems to linger from ancient times. 
Let's git blame this fragment to make sure it's not been added intentionally for any specific reason that is still relevant, try to remove it and see what tests fail.

Reply:
I think the fragility with LIMIT is due to:
 - where it's applied: as rows are read, after filtering, or after aggregation? (should be the latter)
 - how it's passed around to the client and back during paging

Reply:
It's not good to have yet another way to compute the limit.

Maybe you can change the condition in do_get_limit() to more exactly match the circumstances for your fix, and add a comment to explain it.